File: ai-chat-system/backend/src/streamEvent.js
// Unified normalized event shape emitted to clients over SSE.
// Every provider adapter converts its raw stream into these.

export const EventType = Object.freeze({
  START: 'start',
  TOKEN: 'token',
  ERROR: 'error',
  DONE: 'done',
});

export const startEvent = (chatId, messageId, model) => ({
  type: EventType.START, chatId, messageId, model,
});

export const tokenEvent = (delta) => ({ type: EventType.TOKEN, delta });

export const errorEvent = (message, fatal = false) => ({
  type: EventType.ERROR, message, fatal,
});

export const doneEvent = (finishReason = 'stop', usage = null) => ({
  type: EventType.DONE, finishReason, usage,
});

// Serialize a StreamEvent as an SSE frame.
export function sseFrame(event) {
  return `data: ${JSON.stringify(event)}\n\n`;
}

File: ai-chat-system/android/app/src/main/java/com/aichat/app/domain/model/StreamEvent.kt
package com.aichat.app.domain.model

/**
 * The single normalized streaming event the whole app understands.
 * Provider-specific shapes are converted to this by the backend gateway,
 * and the SSE StreamClient parses gateway frames into these.
 */
sealed interface StreamEvent {
    data class Start(val chatId: String, val messageId: String, val model: String) : StreamEvent
    data class Token(val delta: String) : StreamEvent
    data class Error(val message: String, val fatal: Boolean) : StreamEvent
    data class Done(val finishReason: String) : StreamEvent
}
